R/cell_labeling.R

```diff
@@ -0,0 +1,223 @@
+#' @title Calculate normalized expression of a gene set
+#' 
+#' @description Calculate normalized expression for a set of genes in each cell
+#'   from a \code{SingleCellExperiment}, using random sets of similarly
+#'   expressed genes as background to account for cell quality and
+#'   sequencing depth.
+#'
+#' @author Michael Stadler
+#'
+#' @param sce \code{SingleCellExperiment} object.
+#' @param genes \code{character} vector with the genes in the set. Must be a
+#'   subset of \code{rownames(sce)}.
+#' @param expr_values Integer scalar or string indicating which assay of
+#'   \code{sce} contains the expression values.
+#' @param R Integer scalar giving the number of random gene sets to sample
+#'   for normalization.
+#' @param BPPARAM An optional \code{\link[BiocParallel]{BiocParallelParam}}
+#'   instance determining the parallel back-end to be used during evaluation.
+#' 
+#' @return A \code{numeric} vector with normalized gene set scores for each
+#'   cell in \code{sce}.
+#' 
+#' @importFrom BiocParallel SerialParam bplapply
+#' @importFrom SummarizedExperiment assay assayNames
+#' @importFrom matrixStats rowMeans2 rowSds
+#' 
+#' @export
+normGenesetExpression <- function(sce,
+                                  genes,
+                                  expr_values = "logcounts",
+                                  R = 200,
+                                  BPPARAM = SerialParam()) {
+    # pre-flight checks
+    stopifnot(exprs = {
+        is(sce, "SingleCellExperiment")
+        is.character(genes)
+        length(genes) > 1L
+        all(genes %in% rownames(sce))
+        is.numeric(R)
+        length(R) == 1L
+        R > 0
+        inherits(BPPARAM, "BiocParallelParam")
+    })
+    if (is.numeric(expr_values) && length(expr_values) == 1L)
+        expr_values <- assayNames(sce)[expr_values]
+    stopifnot(exprs = {
+        is.character(expr_values)
+        length(expr_values) == 1L
+        expr_values %in% assayNames(sce)
+    })
+    
+    # calculate average expression of selected genes
+    expr <- assay(sce, expr_values)
+    i <- match(genes[genes %in% rownames(expr)], rownames(expr))
+    avgExpr <- rowMeans(expr)
+    val.obs <- colSums(expr[i, , drop = FALSE])
+    
+    # calculate expression in random gene sets
+    val.rand <- do.call(cbind, bplapply(seq_len(R), function(r) {
```

Comment:
Just a thought, and it would make things much more complicated so I'm not sure it's worth it, but would it make sense to exclude highly variable genes from the randomly selected ones, since these can be highly cell type specific and perhaps introduce distortions of the normalized gene set scores (at least if the gene set does not contain a lot of genes)?

Reply:
It seems reasonable to at least allow for that possibility. As highly variable genes are somewhat arbitrarily defined, I will add an option to exclude genes from the sampling, but let the user decide if and which genes to exclude. Does that makes sense?

Reply:
Yes, makes sense!